modules/data-streams/src/main/java/org/elasticsearch/datastreams/UpdateTimeSeriesRangeService.java

```diff
@@ -0,0 +1,191 @@
+                }
+
+                @Override
+                public void clusterStateProcessed(ClusterState oldState, ClusterState newState) {
+                    running.set(false);
+                    onComplete.run();
+                }
+
+                @Override
+                public void onFailure(Exception e) {
+                    running.set(false);
+                    LOGGER.warn("failed to update tsdb data stream end times", e);
+                    onComplete.run();
+                }
+
+            }, ClusterStateTaskExecutor.unbatched());
+        } else {
+            LOGGER.debug("not starting tsdb update task, because another execution is still running");
+        }
+    }
+
+    void setPollInterval(TimeValue newValue) {
+        LOGGER.info(
+            "updating [{}] setting from [{}] to [{}]",
+            DataStreamsPlugin.TIME_SERIES_POLL_INTERVAL.getKey(),
+            pollInterval,
+            newValue
+        );
+        this.pollInterval = newValue;
+    }
+
+    ClusterState updateTimeSeriesTemporalRange(ClusterState current, Instant now) {
+        Metadata.Builder mBuilder = null;
+        for (DataStream dataStream : current.metadata().dataStreams().values()) {
+            if (dataStream.getIndexMode() != IndexMode.TIME_SERIES) {
+                continue;
+            }
+            if (dataStream.isReplicated()) {
+                continue;
+            }
+
+            // getWriteIndex() selects the latest added index:
+            Index head = dataStream.getWriteIndex();
+            IndexMetadata im = current.metadata().getIndexSafe(head);
+            Instant currentEnd = IndexSettings.TIME_SERIES_END_TIME.get(im.getSettings());
+            TimeValue lookAheadTime = IndexSettings.LOOK_AHEAD_TIME.get(im.getSettings());
+            Instant newEnd = now.plus(lookAheadTime.getMillis(), ChronoUnit.MILLIS).plus(pollInterval.getMillis(), ChronoUnit.MILLIS);
+            if (newEnd.isAfter(currentEnd)) {
+                try {
+                    Settings settings = Settings.builder()
+                        .put(IndexSettings.TIME_SERIES_END_TIME.getKey(), DEFAULT_DATE_TIME_FORMATTER.format(newEnd))
+                        .build();
+                    LOGGER.debug(
+                        "updating [{}] setting from [{}] to [{}] for data stream [{}]",
+                        IndexSettings.TIME_SERIES_END_TIME.getKey(),
+                        currentEnd,
+                        newEnd,
+                        dataStream.getName()
+                    );
+                    if (mBuilder == null) {
+                        mBuilder = Metadata.builder(current.metadata());
+                    }
+                    mBuilder.updateSettings(settings, head.getName());
+                    Metadata.Builder finalMBuilder = mBuilder;
+                    // Verify that all temporal ranges of each backing index is still valid:
+                    dataStream.validate(finalMBuilder::get);
```

Comment:
Sorry, I don't see what you mean by not being able to pass `mBuilder::get` into the validate method here. At this point the mBuilder is guaranteed to be initialized (since we initialize it if null a couple of lines above), and passing `mBuilder::get` doesn't cause any compilation failure. I'd like to understand better, what am I missing?

Reply:
I was pretty sure that when I made this change, then IntelliJ reported a compile error (can't use a variable that isn't final or effectively final in a lambda). But not anymore, so I will remove this `finalMBuilder ` variable.

Reply:
Implemented via: dd6c400b230d4c1fc8df65760d3c7eb3c01d6192